src/utils/cid/mod.rs

```diff
@@ -28,30 +28,47 @@ impl CidCborExt for Cid {}
 
 pub const BLAKE2B256_SIZE: usize = 32;
 
-// `CidVariant` is an enumeration of known CID types that are used in the Filecoin blockchain. CIDs
-// contain a significant amount of static data (such as version, codec, hash identifier, hash
-// length). This static data represented by a single tag in the enum.
-//
-// Nearly all Filecoin CIDs are V1, DagCbor encoded, and hashed with Blake2b256 (which has a hash
-// length of 256bits). Naively representing such a CID requires 96 bytes but `CidVariant` does it in
-// only 40 bytes. If other types of CID become popular, they can be added to the CidVariant
-// structure.
+/// `CidVariant` is an enumeration of known CID types that are used in the Filecoin blockchain. CIDs
+/// contain a significant amount of static data (such as version, codec, hash identifier, hash
+/// length). This static data represented by a single tag in the `enum`.
+///
+/// Nearly all Filecoin CIDs are `V1`,`DagCbor` encoded, and hashed with `Blake2b256` (which has a hash
+/// length of 256 bits). Naively representing such a CID requires 96 bytes but `CidVariant` does it in
+/// only 40 bytes. If other types of CID become popular, they can be added to the `CidVariant`
+/// structure.
+///
+/// The `Generic` variant is used for CIDs that do not fit into the other variants.
+/// These variants are used for optimizing storage of CIDs in the `FrozenCids` structure.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
 pub enum CidVariant {
```

Comment:
```suggestion
pub enum SmallCid {
```

Reply:
should also `#[derive(Arbitrary)]`

Reply:
I think `CidVariant` is more accurate--only the `V1DagCborBlake2b` variant is actually smaller, and renaming that variant to `SmallCid` would lose descriptiveness as well. For the `Arbitrary` discussion, see https://github.com/ChainSafe/forest/pull/3275#discussion_r1288798090.

Reply:
Not going to hold up the PR on this, but to explain my thinking.

`CidVariant` is a bit like `CidKind` - it sounds like a classification of a CID. Think of [`std::io::ErrorKind`](https://doc.rust-lang.org/stable/std/io/enum.ErrorKind.html).
We don't want that - we're abstracting over different _storage methods_ of CIDs.

See other similar crates:
- https://crates.io/crates/smol_str
- https://docs.rs/smallvec/latest/smallvec/

Reply:
Also, I think this is the layout we want:
```rust
/// Save space on common CID types
pub struct SmallCid(SmallCidInner);

impl SmallCid {
    pub fn cid(&self) -> Cid { ... }
}

impl From<Cid> for SmallCid { ... }
impl From<SmallCid> for Cid { ... }

enum SmallCidInner {
    DagCborV1([u8; 32]),
    Heap(Box<Cid>)
}
```

Users shouldn't be able to see the enum variant